Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+gap between Ceph and Kubernetes, putting it in a unique domain with its own best practices to
+follow. Therefore, this document specifically covers best practices for running Ceph on Kubernetes
+with Rook. Because Rook augments on top of Kubernetes, it has different ways of meeting both Ceph
+and Kubernetes best practices. This is in comparison to the bare metal versions of each. Out of the
+box, Rook is predominantly a default Ceph cluster. The Ceph cluster needs tuning to meet user
+workloads, and Rook does not absolve the user from planning out their production storage cluster
+beforehand.
+
+For the purpose of this document, we will consider two simplified use cases to help us make informed
+decisions about Rook and Ceph:
+ * Co-located: User applications co-exist on nodes running Ceph
+ * Disaggregated: Ceph nodes are totally separated from user applications
+
+
+## General Best Practices
+
+This chapter provides an outline of a series of generalized recommendations for best practices:
+* Ceph monitors are more stable on fast storage (SSD-class or better) according to Ceph best
+  practices. In Rook, this means that the `dataDirHostPath` location in the `cluster.yaml` should be
+  backed by SSD or better on MON nodes.
+* Raise the Rook log level to `DEBUG` for initial deployment and for upgrades, as it will help with
+  debugging problems that are more likely to occur at those times.</br>
+  Ensure that the `ROOK_LOG_LEVEL` in operator.yaml equals `debug`.
+* The Kubernetes CSI driver is the preferred default, but ensure in `operator.yaml` that
+  `ROOK_ENABLE_FLEX_DRIVER` remains set to `false`. This is because the FlexVolume driver is in
+  sustaining mode, is not getting non-priority bug fixes, and will soon be deprecated.
+* Ceph’s placement group (PG) auto-scaler module makes it unnecessary to manually manage PGs. We
+  recommend you always set this to `enabled`, unless you have some need to manage PGs manually. In
+  `cluster.yaml`, enable the `pg_autoscaler` MGR module.
+* Rook has the capability to auto-remove Deployments for OSDs which are kicked out of a Ceph
+  cluster. This is enabled by: `removeOSDSIfOutAndSafeToRemove: true`. This means there is less user
+  OSD maintenance and no need to delete Deployments for OSDs that have been kicked out. Rook will
+  automatically clean up the cluster by removing OSD Pods if the OSDs are no longer holding Ceph
+  data. However, keep in mind that this can reduce the visibility of failures from Kubernetes Pod
+  and Pod Controller views. You can optionally set `removeOSDSIfOutAndSafeToRemove` to `false` if
+  need be, such as if a Kubernetes administrator wants to see disk failures via a Pod overview.
+* Configure Ceph using the central configuration database when possible. This allows for more
+  runtime configuration flexibility. Do this using `ceph config set` commands from Rook's toolbox.
+  Only use Rook’s provided ceph.conf override ConfigMap when it is required.
```

Comment:
What is the best practice for when the ceph.conf is required? For example, in these situations when the setting is:
- Not available via `ceph config set`
- Needed at initial cluster setup
- The toolbox is not available to run the `ceph config set`

Reply:
TBH because using the ConfigMap means restarting the daemons to apply the configuration I see very little value in it unless we want fully automate deployments and apply a config during the initial cluster setup.
So I'd personally only document the initial cluster setup.

If the toolbox is not available the admin can still use the operator pod (with a bit of trick like ` export CEPH_ARGS="-c /var/lib/rook/rook-ceph/rook-ceph.config"`)